sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -128,30 +136,32 @@ private[sql] object PartitioningUtils {
   private[sql] def parsePartition(
       path: Path,
       defaultPartitionName: String,
-      typeInference: Boolean): Option[PartitionValues] = {
+      typeInference: Boolean): (Option[PartitionValues], Option[Path]) = {
```

Comment:
Or change the `case class PartitionValues(columnNames: Seq[String], literals: Seq[Literal])` as `case class PartitionValues(path: String, columnNames: Seq[String], literals: Seq[Literal])`?

Then the code probably much simple and readable.